Ast opencl fixes

sakehl opened the pull request:
Mainly added support for rewriting nested foralls, which are automatically generated by GPU kernels and par blocks.
This is done in the pass [SimplifyNestedQuantifiers.scala](https://github.com/utwente-fmt/vercors/compare/ast...sakehl:vercors:ast-opencl-fixes#diff-d88a9200060089c6e668027c093fd6acdf0d3959e2bda429e994b04bda1ad2d2).

This pass relies on [ExpressionEqualityCheck.scala](https://github.com/utwente-fmt/vercors/compare/ast...sakehl:vercors:ast-opencl-fixes#diff-1a3f5ada4669de4f01d73f3d646ba1ebc9336c15983024f57fd7b204f19fcc39), to compare expressions on equality, whilst getting information like `require blockSize == 32` from annotations.

There are some changes in how a GPU kernel is encoded ([LangCToCol.scala](https://github.com/utwente-fmt/vercors/compare/ast...sakehl:vercors:ast-opencl-fixes#diff-22aa2036ea2d11dfc54ccc748ab0c2b081cebc051a76b80ffca2832f0ec3a556)) and how par blocks are encoded ([ParBlockEncoder.scala](https://github.com/utwente-fmt/vercors/compare/ast...sakehl:vercors:ast-opencl-fixes#diff-30317b8bbd865428e4b19fb47fe9eccada34d3e1b232ea4c63de25851f25605c)), to make sure that par blocks and GPU kernels require to have at least one thread running (therefor, we don't have to account for "void" runs of GPU kernels). Which was needed to gather information, which is needed in the expressions equality check, which was used in the simplified nested foralls.

Missing are some examples and unit tests still.

Also a lot of other minor changes are present:
* added bool type to opencl and cuda headerse
* After quantifiers are rewritten in simplify relations, to original name is kept (or at least we try to)
* C arrays and variables will keep their internal name
* Some things I probably forgot I changed.

niomaster reviewed (commented):
It's taking me a bit longer than expected to review everything, so here is a partial review so it doesn't come late in one batch :)

sakehl commented on src/main/java/vct/main/stages/Transformation.scala:
```diff
@@ -149,14 +149,21 @@ case class SilverTransformation
+    ) ++ simplifyAfterRelations ++ Seq(
+
+    // Encode proof helpers
+    EncodeProofHelpers,
```
Encode proof helpers changed FramedProofs requires and ensures to be changed to loop invariants with `once ==> ...` and `!once ==> ..`, which made SimplifyNestedQuantifiers not work anymore, since it did not allow to extract things like `threadBlockSize == 32`. (since we now have `once ==> threadBlockSize == 32`). 

I could change it I suppose to work with that, but it makes the pass more difficult with more bookholding.

Similar for `EncodeTryThrowSignals`. What would you suggest?

sakehl commented on src/main/java/vct/col/newrewrite/ParBlockEncoder.scala:
```diff
@@ -104,31 +104,72 @@ case class ParBlockEncoder[Pre <: Generation]() extends Rewriter[Pre] {
+      quantify(block, block.context_everywhere &* block.ensures, nonEmpty)
+  }
+
+  def constantExpression(e: Expr[_]): Boolean = e match {
+    case _: Constant[_, _] => true
+    case op: BinExpr[_] => constantExpression(op.left) && constantExpression(op.right)
+    /* TODO: This is the hard part, do we do an analysis on locals to see if they are never changed in the body of a a
+     *  parallel block? We need this, otherwise we can never hope to rewrite nested foralls, since the bounds will
+     *  depend on the hi_var en var_low values. And on that the nested quantifier pass will get stuck.
+     */
+    case _: Local[_] => false
+    case _ => false
```
Suggestions for this?

sakehl commented on src/main/java/vct/col/newrewrite/ParBlockEncoder.scala:
```diff
@@ -104,57 +104,91 @@ case class ParBlockEncoder[Pre <: Generation]() extends Rewriter[Pre] {
-      })(region.o)
+      iters.foreach { v =>
+        rangeValues(v.variable) = (dispatch(v.from), dispatch(v.to))
+      }
```
I added this back to be separately evaluated. But this breaks SimplifyNestedQuantifiers (always) for GPU programs.

I would want to fix this by determining when an expression is constant (and thus doesn't change), where this approach is correct. Thus implementing `constantExpression` (see lin 164). But then we need to argue if certain variables never change in the body of some code. Suggestions?

sakehl replied:
In my last commit, I added a solution to this but it would be nice to see if you agree with this.

niomaster replied:
yes, this is a good approach. I'll likely extract this and make it more complete in the future.

sakehl commented:
> It's taking me a bit longer than expected to review everything, so here is a partial review so it doesn't come late in one batch :)

I fixed the changes you suggested. Only 2 things remain open, see above. Also added barriers with memory fences

niomaster commented on parsers/src/main/java/vct/parsers/ColCParser.scala:
```diff
@@ -24,7 +24,8 @@ case class ColCParser(override val originProvider: OriginProvider,
                       otherIncludes: Seq[Path],
-                      defines: Map[String, String]) extends Parser(originProvider, blameProvider) with LazyLogging {
+                      defines: Map[String, String],
+                      language: Language) extends Parser(originProvider, blameProvider) with LazyLogging {
```
Let's not add language information to transformations. I see that you primarily use it to interpret memory locality based on the language, but rather the AST should speak for itself.
